challenges/systems/challenge-2.md

```diff
@@ -0,0 +1,296 @@
+  questions in Slack and sharing a design document that is ready to be reviewed.
+
+Once the design document has been approved by two reviewers, move onto
+implementation.
+
+Split your code submission into roughly 3-5 Pull Requests to give the team an
+opportunity to review your code and provide feedback. Feel free to merge each
+PR after you have two approvals.
+
+Our team will do their best to provide a high quality review of the submitted
+Pull Requests in a reasonable time frame. You are spending your time on this,
+we are going to contribute our time too.
+
+After the final submission, the interview team will assemble and vote using a
+"+1, -2" anonymous voting system: +1 is submitted whenever a team member
+accepts the submission, -2 otherwise.
+
+In case of a positive result, we will connect you to our HR and recruiting
+teams, who will work out the details and present an offer.
+
+In case of a negative score result, hiring manager will contact you and share a
+list of the key observations from the team that affected the result.
+
+# Requirements
+
+## Level 1
+
+### Library
+
+* Implement a simple request forwarder to forward requests to an upstream
+
+### Server
+
+* Secure communication between client and server using TLS and token
+  authentication
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using library
+
+## Level 2
+
+### Library
+
+* Implement a round-robin request forwarder to forward requests to
+  multiple upstreams
+
+### Server
+
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using library
+
+## Level 3
+
+### Library
+
+* Track number of connections per upstreams and implement a least connections
+  request forwarder
+
+### Server
+
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using library
+
+## Level 4
+
+### Library
+
+* Track number of connections per upstreams and implement a least connections
+  request forwarder
+* Track number of connections per client and implement a rate limiting request
+  forwarder
+
+### Server
+
+* Use mTLS authentication to have the server verify identity of the client and
+  client of the server
+* Develop a simple authorization scheme that defines what upstreams are
+  available to which clients; this scheme can be statically defined in code
+* Accept and forward requests to upstreams using library
+
+## Level 5
```

Comment:
What do you guys think of not repeating the same details for every level and saying something like "everything in level 4 plus..." instead. The first thing I did for the backend challenge was to highlight what was new for each level.